pkg/resource/stack/secrets.go

```diff
@@ -125,132 +123,404 @@ func (csm *cachingSecretsManager) State() json.RawMessage {
+
+// BeginEncryptionBatch returns a new BatchEncrypter and CompleteCrypterBatch function.
+// The BatchEncrypter allows encryption operations to be batched to avoid round-trips for when the underlying encrypter
+// requires a network call. Encryptions to a secret are enqueued and processed in a batch request either when the queue
+// is full or when the wrapping transaction is committed.
+//
+// The CompleteCrypterBatch function must be called to ensure that all enqueued encryption operations are processed.
 //
-// Note that this is intentionally separate from cachingCrypter. A cachingCrypter is intended to prevent repeated
-// encryption of secrets when the same snapshot is repeatedly serialized over the lifetime of an update, and
-// therefore keys on the identity of the secret value itself. A mapDecrypter is intended to allow the deserializer
-// to decrypt secrets up-front and prevent repeated calls to decrypt within the context of a single deserialization,
-// and cannot key off of secret identity because secrets do not exist when the cache is initialized.
-type mapDecrypter struct {
-	decrypter config.Decrypter
-	cache     map[string]string
+//	batchEncrypter, completeCrypterBatch := BeginEncryptionBatch(encrypter)
+//	defer completeCrypterBatch(ctx)
+//	SerializeSecrets(ctx, batchEncrypter, secrets)
+func BeginEncryptionBatch(encrypter config.Encrypter) (BatchEncrypter, CompleteCrypterBatch) {
+	return BeginEncryptionBatchWithCache(encrypter, NewSecretCache())
 }
 
-func newMapDecrypter(decrypter config.Decrypter, cache map[string]string) config.Decrypter {
-	return &mapDecrypter{decrypter: decrypter, cache: cache}
+// BeginEncryptionBatchWithCache returns a new BatchEncrypter and CompleteCrypterBatch function with a custom cache.
+// The BatchEncrypter allows encryption operations to be batched to avoid round-trips for when the underlying encrypter
+// requires a network call. Encryptions to a secret are enqueued and processed in a batch request either when the queue
+// is full or when the wrapping transaction is committed. If the cache has an entry for a secret and the plaintext has
+// not changed, the previous ciphertext is used immediately and not enqueued for the batch operation. Results are
+// also written to the provided cache.
+//
+// The CompleteCrypterBatch function must be called to ensure that all enqueued encryption operations are processed.
+//
+//	batchEncrypter, completeCrypterBatch := BeginEncryptionBatch(encrypter)
+//	defer completeCrypterBatch(ctx)
+//	SerializeSecrets(ctx, batchEncrypter, secrets)
+func BeginEncryptionBatchWithCache(
+	encrypter config.Encrypter, cache SecretCache,
+) (BatchEncrypter, CompleteCrypterBatch) {
+	return beginEncryptionBatch(encrypter, cache, DefaultMaxBatchEncryptCount)
 }
 
-func (c *mapDecrypter) DecryptValue(ctx context.Context, ciphertext string) (string, error) {
-	if plaintext, ok := c.cache[ciphertext]; ok {
-		return plaintext, nil
+func beginEncryptionBatch(
```

Comment:
So to my mind, these three functions are the factories where we should be checking e.g. is the env var for caching disabled and if so, not decorating with a caching layer. Equally, I think their names should be more like `NewBatchingCachingEncrypter` or the like -- they don't actually start or enqueue anything, I don't think?

Reply:
I've now updated this to check the env in the secret cache construction function instead, and the non caching constructors for the batching crypters always use the nullSecretsCache too.